integration/src/main/java/scripts/testscripts/CloneWorkspace.java

```diff
@@ -0,0 +1,411 @@
+import bio.terra.workspace.api.ControlledGcpResourceApi;
+import bio.terra.workspace.api.ReferencedGcpResourceApi;
+import bio.terra.workspace.api.WorkspaceApi;
+import bio.terra.workspace.client.ApiClient;
+import bio.terra.workspace.model.CloneResourceResult;
+import bio.terra.workspace.model.CloneWorkspaceRequest;
+import bio.terra.workspace.model.CloneWorkspaceResult;
+import bio.terra.workspace.model.CloningInstructionsEnum;
+import bio.terra.workspace.model.ControlledResourceIamRole;
+import bio.terra.workspace.model.CreatedControlledGcpGcsBucket;
+import bio.terra.workspace.model.GcpBigQueryDatasetResource;
+import bio.terra.workspace.model.GcpGcsBucketResource;
+import bio.terra.workspace.model.GrantRoleRequestBody;
+import bio.terra.workspace.model.IamRole;
+import bio.terra.workspace.model.PrivateResourceIamRoles;
+import bio.terra.workspace.model.ResourceCloneDetails;
+import bio.terra.workspace.model.ResourceType;
+import bio.terra.workspace.model.StewardshipType;
+import bio.terra.workspace.model.WorkspaceDescription;
+import com.google.cloud.bigquery.BigQuery;
+import com.google.cloud.bigquery.QueryJobConfiguration;
+import com.google.cloud.bigquery.TableResult;
+import com.google.cloud.storage.Blob;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.common.collect.ImmutableList;
+import java.io.IOException;
+import java.time.Duration;
+import java.util.List;
+import java.util.UUID;
+import java.util.stream.StreamSupport;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scripts.utils.ClientTestUtils;
+import scripts.utils.CloudContextMaker;
+import scripts.utils.DataRepoTestScriptBase;
+import scripts.utils.ResourceMaker;
+import scripts.utils.WorkspaceAllocateTestScriptBase;
+
+public class CloneWorkspace extends WorkspaceAllocateTestScriptBase {
+  private static final Logger logger = LoggerFactory.getLogger(CloneWorkspace.class);
+  private ControlledGcpResourceApi cloningUserResourceApi;
+  private CreatedControlledGcpGcsBucket copyDefinitionSourceBucket;
+  private CreatedControlledGcpGcsBucket privateSourceBucket;
+  private CreatedControlledGcpGcsBucket sharedCopyNothingSourceBucket;
+  private CreatedControlledGcpGcsBucket sharedSourceBucket;
+  private GcpBigQueryDatasetResource sourceDatasetReference;
+  private GcpBigQueryDatasetResource copyDefinitionDataset;
+  private GcpBigQueryDatasetResource copyResourceDataset;
+  private GcpBigQueryDatasetResource privateDataset;
+  private GcpGcsBucketResource sourceBucketReference;
+  private String copyDefinitionDatasetName;
+  private String copyResourceDatasetName;
+  private String nameSuffix;
+  private String privateDatasetName;
+  private String sharedBucketSourceResourceName;
+  private String sourceProjectId;
+  private TestUserSpecification sourceOwnerUser;
```

Comment:
nit: make `sourceOwnerUser` local to `doSetup`? Otherwise it's shadowed in `doUserJourney`. Pretty minor since they're the same user anyways

Reply:
good eye